### ISSUE TITLE
allow null in relationship data resource linkage list

### ISSUE THREAD

**Issue by shaunc** [json-api/json-api#1240]
It seems to me that the current specification has an inconsistency. The resource linkage may support ordered relationships, and an ordered relationship may have missing elements. Thus, the resource linkage list should allow null elements to represent to-many ordered relationships in which some elements are missing.

**Comment by beauby** [comment]
Could you provide an example use-case with a corresponding payload?

**Comment by shaunc** [comment]
Here is an example of a document of a resource with relationship "foo" to "foos", in which 0th and 2nd element correspond to a resource, and 1st is empty:

```json
{
  "data": {
  "id": "some-id",
  "type": "some-types",
    "relationships": {
      "foo": {
        "data": [{"id": "1", "type": "foos"}, null, {"id": "10", "type": "foos"}]
      }
    }
  }
}
```

**Comment by ethanresnick** [comment]
This would be a breaking change, in the sense that it would break existing consumers—clients or servers—reading documents produced with this new payload. Whether that's a breaking change we *could* permit is a [bit unclear](https://github.com/json-api/json-api/issues/949) (we need to better define what we mean by "additive only" in future specs). Sill, even if such a change is possible, I'm not sure that the breakage and extra complexity (there are already a lot of values possible in data) would be worth it—especially since the idea that a relationship is ordered is already optional/at the periphery of the spec. 

Therefore, my recommendation would be to handle this with resource identifier object meta, like so:

```
{
  "data": {
  "id": "some-id",
  "type": "some-types",
    "relationships": {
      "foo": {
        "data": [
          // note how index 1 is missing
          {"id": "1", "type": "foos", "meta": { "index": 0 } }, 
          {"id": "10", "type": "foos", "meta": { "index": 2 } }
        ]
      }
    }
  }
}
```

**Comment by shaunc** [comment]
A related issue is POST requests to a relationship. An ordered relationship could have the same element more than once. Would a server be compliant if it used the index field in meta to denote which element of the ordered relationship it should replace (optionally, in this case, allowing more than one of the same id/type)? If so, IMO that should be reflected in the specification explicitly, as otherwise it would seem that a server supporting an ordered, non-unique relationship would be not in compliance.

**Afterthought:** Perhaps a conservative extension would be to take the "index" parameter out of meta and make it an optional part of the relationship data itself (alongside id and type). Then "meta" wouldn't be constrained by the protocol and "index" in meta could mean whatever it the implementors want from it.

**Comment by jelhan** [comment]
I don't agree that supporting ordered relationships implies that we also need to support 1) missing / empty elements in an ordered list and 2) parallel relationships.

There is an open PR to clarify support for parallel relationships: #1677 Referencing the same resource twice from the same relationship without intermediate resource should be discouraged in my perspective.

Explicitly marking elements as missing / empty in an ordered list, is an edge case feature in my perspective. I don't see a need supporting it in base specification. I think an extension should be used if this is needed.
